packages/serverless-orchestration/src/ServerlessHub.js

```diff
@@ -95,17 +95,39 @@ hub.post("/", async (req, res) => {
       // Check if bot is running on a non-default chain, and fetch last block number seen on this or the default chain.
       const [botWeb3, spokeCustomNodeUrl] = _getWeb3AndUrlForBot(configObject[botName]);
       const chainId = await _getChainId(botWeb3);
-      nodeUrlToChainIdCache[spokeCustomNodeUrl] = chainId;
-
-      // If we've seen this chain ID already we can skip it:
+      // If we've seen this chain ID already we can skip it.
       if (blockNumbersForChain[chainId]) continue;
 
-      // Fetch last seen block for this chain:
-      let lastQueriedBlockNumber = await _getLastQueriedBlockNumber(req.body.configFile, chainId, logger);
+      nodeUrlToChainIdCache[spokeCustomNodeUrl] = chainId;
+
+      // If STORE_MULTI_CHAIN_BLOCK_NUMBERS is set then this bot requires to know a number of last seen blocks across
+      // a set of chainIds. Construct a batch promise to evaluate the latest block number for each chainId.
+      if (configObject[botName]?.environmentVariables?.STORE_MULTI_CHAIN_BLOCK_NUMBERS) {
+        const multiChainIds = configObject[botName]?.environmentVariables?.STORE_MULTI_CHAIN_BLOCK_NUMBERS;
+        let promises = [];
+        for (const chainId of multiChainIds) {
+          promises.push(
+            _getLastQueriedBlockNumber(req.body.configFile, chainId, logger),
+            _getBlockNumberOnChainIdMultiChain(configObject[botName], chainId)
+          );
+        }
+        let results = await Promise.all(promises);
+        results.forEach((_, index) => {
```

Comment:
nit: I think we could drop some of this even odd logic if we were to use a map. (you would lose the parallelization of the inner calls, but I don't think that's a huge deal, and could be compensated for if you did an inner Promise.all)

```ts
const blockNumbersForChain = await Promise.all(multiChainIds.map(async (chainId) => ({
  lastQueriedBlockNumber: await _getLastQueriedBlockNumber(req.body.configFile, chainId, logger),
  latestBlockNumber: await _getBlockNumberOnChainIdMultiChain(configObject[botName], chainId),
})));
```

Not necessary to do now. Could be done in follow up.

Reply:
I originally had it exactly like this! I changed it to this structure to be more parallel. the problem is that there are 5 chains that would have some level of blocking if done the way you have it now. there will be 5 pairs of 2 Promise.all calls. my structure has 1 set of 10 calls, all at once.